File: eeg_autocorr_fftw.m
function [ac2,resamp] = eeg_autocorr_fftw(icaact, trials, srate, pnts, pct_data)
if ~exist('pct_data', 'var') || isempty(pct_data)
    pct_datas = 100;
end

nfft = 2^nextpow2(2*pnts-1);

% calc autocorrelation
fftw('planner', 'hybrid');
ac = zeros(size(icaact, 1), nfft, trials);

for it = 1:size(icaact, 1)
    X = fft(icaact(it, :, :), nfft, 2);
    %if it == 1
    %    ac2 = X(:,:,:);
    %end
    ac(it, :, :) = abs(X).^2;
end
ac2 = ac(:,:,:);
ac = ifft(mean(ac, 3), [], 2);

if pnts < srate
    ac = [ac(:, 1:pnts) zeros(size(ac, 1), srate - pnts + 1)];
else
    ac = ac(:, 1:srate + 1);
end

% normalize by 0-tap autocorrelation
ac = bsxfun(@rdivide, ac(:, 1:srate + 1, :), ac(:, 1)); 

% resample to 1 second at 100 samples/sec
resamp = resample(ac', 100, srate)';
resamp(:, 1) = [];

File: eeg_autocorr_fftw.py
import math
import numpy as np
import scipy.signal as ss
from scipy.fft import fft, ifft


def eeg_autocorr_fftw(icaact: np.array, trials: int, srate: float, pnts: int, pct_data: int = 100) -> np.array:
    """
    Generates autocorrelation features for ICLabel.

    Args:
        icaact (np.array): ICA activation waveforms
        trials (int): number of trials
        srate (float): sample rate
        pnts (int): number of points
        pct_data (int, optional): [description]. Defaults to 100.

    Returns:
        np.array: autocorrelation feature
    """
    nfft = 2**(math.ceil(math.log2(abs(2*pnts - 1))))
    ac = np.zeros((len(icaact), nfft, trials), dtype=np.float64)
    
    for it in range(len(icaact)):
        X = fft(icaact[it:it+1,:,:], n = nfft, axis = 1)
        ac[it:it+1,:,:] = np.power(np.abs(X),2)
    
    ac = np.abs(ifft(np.mean(ac, axis=2), n=None, axis=1)) # ifft
    
    if pnts < srate:
        ac = np.hstack((ac[:,0:pnts], np.zeros((len(ac), srate - pnts + 1))))
    else:
        ac = ac[:,0:srate+1]

    ac = ac[:,0:srate+1] / ac[:,1][:,None]

    resamp = ss.resample_poly(ac.T, 100, srate).T

    return resamp[:,1:]
    